src/box/memtx_sort_data.c

```diff
@@ -0,0 +1,689 @@
+#define MH_SOURCE
+#include "salad/mhash.h"
+
+/** The sort data file writer context. */
+struct memtx_sort_data_writer {
+	/** The sort data file pointer. */
+	FILE *fp;
+	/** The sort data file name. */
+	char filename[PATH_MAX];
+	/** The information about the sort data entries. */
+	struct mh_memtx_sort_data_writer_entries_t *entries;
+	/** The currently handled sort data entry. */
+	struct memtx_sort_data_writer_entry *curr_entry;
+	/** The offset to the next index entry in the file header. */
+	long next_header_entry_offset;
+};
+
+/** The sort data reader context. */
+struct memtx_sort_data_reader {
+	/** The sort data file pointer. */
+	FILE *fp;
+	/** The sort data file name. */
+	char filename[PATH_MAX];
+	/** The information about the sort data entries. */
+	struct mh_memtx_sort_data_entries_t *entries;
+	/** The currently handled sort data entry, NULL if space is skipped. */
+	struct memtx_sort_data_reader_entry *curr_entry;
+	/** The old to new tuple address map used on recovery of a space. */
+	struct mh_ptrptr_t *old2new;
+	/** The buffer to use on fread. */
+	void *buffer;
+};
+
+/* {{{ Utilities **************************************************************/
+
+/** Sort data file header format strings. */
+static const char *ENTRIES_FMT = "Entries: %010u\n";
+static const char *ENTRY_FMT = "%010u/%010u: %016lx, %016lx, %20li\n";
+
+/**
+ * Returns the sort data file name based on the corresponding @a snap_filename.
+ */
+static const char *
+memtx_sort_data_filename(const char *snap_filename)
+{
+	char *snap_ext = strrchr(snap_filename, '.');
+	assert(snap_ext != NULL);
+	if (strcmp(snap_ext, ".inprogress") == 0) {
+		snap_ext -= strlen(".snap");
+		assert(snap_ext > snap_filename);
+	}
+	assert(strcmp(snap_ext, ".snap") == 0 ||
+	       strcmp(snap_ext, ".snap.inprogress") == 0);
+	int snap_filename_noext_len = snap_ext - snap_filename;
+	return tt_snprintf(PATH_MAX, "%.*s.sortdata",
+			   snap_filename_noext_len, snap_filename);
+}
+
+/** Seek to the specified offset in the sort data file. */
+static int
+seek(struct memtx_sort_data_writer *writer, long offset, int whence)
+{
+	if (fseek(writer->fp, offset, whence) != 0) {
+		diag_set(SystemError, "%s: failed to fseek in the "
+			 "sort data file", writer->filename);
+		return -1;
+	}
+	return 0;
+}
+
+/** Get the current sort data file offset. */
+static int
+get_offset(struct memtx_sort_data_writer *writer, long *offset)
+{
+	*offset = ftell(writer->fp);
+	if (*offset == -1) {
+		diag_set(SystemError, "%s: failed to ftell on the "
+			 "sort data file", writer->filename);
+		return -1;
+	}
+	return 0;
+}
+
+/** Write the formatted string into the sort data file. */
+static int
+print(struct memtx_sort_data_writer *writer, const char *fmt, ...)
+{
+	va_list args;
+	va_start(args, fmt);
+	if (vfprintf(writer->fp, fmt, args) < 0) {
+		diag_set(SystemError, "%s: failed to write the sort data file",
+			 writer->filename);
+		return -1;
+	}
+	va_end(args);
+	return 0;
```

Comment:
Nit: it's a good practice not to call functions with trivial names in pure C world to avoid symbol conflicts. Let's rename them. Also, as another reviewer already said, such short names make people think about syscalls or libc.